pageserver/src/relish_storage.rs

```diff
@@ -0,0 +1,171 @@
+}
+
+impl std::error::Error for RelishStorageError {}
+
+impl From<S3Error> for RelishStorageError {
+    fn from(e: S3Error) -> Self {
+        Self::S3(e)
+    }
+}
+
+impl From<LocalFsError> for RelishStorageError {
+    fn from(e: LocalFsError) -> Self {
+        Self::Local(e)
+    }
+}
+
+impl GenericRelishStorage {
+    pub fn std_local_fs(root: PathBuf) -> Result<Self, RelishStorageError> {
+        Ok(Self::StdLocalFs(StdLocalFs::new(root)?))
+    }
+
+    pub fn rust_s3(
+        config: &S3Config,
+        tenant_id: Option<ZTenantId>,
+    ) -> Result<Self, RelishStorageError> {
+        Ok(Self::RustS3(RustS3::new(config, tenant_id)?))
+    }
+}
+
+impl RelishStorage for GenericRelishStorage {
+    type Error = RelishStorageError;
+
+    fn list_relishes(&self) -> Result<Vec<RelishStoragePath>, Self::Error> {
+        Ok(match self {
+            GenericRelishStorage::StdLocalFs(std_local_fs) => std_local_fs.list_relishes()?,
+            GenericRelishStorage::RustS3(rust_s3) => rust_s3.list_relishes()?,
+        })
+    }
+
+    fn download_relish(&self, path: &RelishStoragePath) -> Result<Vec<u8>, Self::Error> {
+        Ok(match self {
+            GenericRelishStorage::StdLocalFs(std_local_fs) => std_local_fs.download_relish(path)?,
+            GenericRelishStorage::RustS3(rust_s3) => rust_s3.download_relish(path)?,
+        })
+    }
+
+    fn delete_relish(&self, path: &RelishStoragePath) -> Result<(), Self::Error> {
+        Ok(match self {
+            GenericRelishStorage::StdLocalFs(std_local_fs) => std_local_fs.delete_relish(path)?,
+            GenericRelishStorage::RustS3(rust_s3) => rust_s3.delete_relish(path)?,
+        })
+    }
+
+    fn store_relish(
+        &self,
+        relish_timeline_id: ZTimelineId,
+        relish_name: &str,
+        contents: Vec<u8>,
+    ) -> Result<RelishStoragePath, Self::Error> {
+        Ok(match self {
+            GenericRelishStorage::StdLocalFs(std_local_fs) => {
+                std_local_fs.store_relish(relish_timeline_id, relish_name, contents)?
+            }
+            GenericRelishStorage::RustS3(rust_s3) => {
+                rust_s3.store_relish(relish_timeline_id, relish_name, contents)?
+            }
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct S3Error(String);
+
+impl std::fmt::Display for S3Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "s3 error: {}", self.0)
+    }
+}
+
+impl std::error::Error for S3Error {}
+
+#[derive(Debug)]
+pub struct LocalFsError(String);
+
+impl std::fmt::Display for LocalFsError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "s3 error: {}", self.0)
+    }
+}
+
+impl std::error::Error for LocalFsError {}
+
+impl From<std::io::Error> for LocalFsError {
+    fn from(e: std::io::Error) -> Self {
+        Self(e.to_string())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RelishStoragePath {
```

Comment:
I agree that we need to simplify the structure internals, especially due to the fact that no extensions of whatnot is needed now, but don't agree to use `PathBuf` everywhere, since Windows PathBuf might bring surprises when used in s3:

<img width="601" alt="in" src="https://user-images.githubusercontent.com/2690773/133051234-63914f72-207f-4026-aba3-3573ccd1afcc.png">
<img width="613" alt="out" src="https://user-images.githubusercontent.com/2690773/133051241-a4f37c31-4c48-441f-99f6-661ac546312f.png">

So most probably, some other trait can be used, if we want to abstract it.
I'll think of options.